File: framework/src/main/java/com/thornbirds/frameworkext/component/route/RouteComponentController.java
package com.thornbirds.frameworkext.component.route;

import android.util.Log;

import androidx.annotation.Nullable;

import com.thornbirds.component.ComponentController;
import com.thornbirds.component.IEventController;

/**
 * Created by yangli on 2017/11/19.
 */
public abstract class RouteComponentController<T extends IRouter> extends ComponentController {

    public static final int STATE_IDLE = 0;
    public static final int STATE_CREATED = 1;
    public static final int STATE_STARTED = 2;
    public static final int STATE_STOPPED = 3;
    public static final int STATE_DESTROYED = 4;

    @Nullable
    public abstract T getRouter();

    @Nullable
    public abstract T getParentRouter();

    public abstract boolean exitRouter();

    private int mState = STATE_IDLE;

    public final int getState() {
        return mState;
    }

    protected abstract void onCreate();

    protected abstract void onStart();

    protected abstract void onStop();

    protected abstract void onDestroy();

    protected abstract boolean onBackPressed();

    public final void performCreate() {
        if (mState == STATE_CREATED) {
            return;
        }
        if (mState == STATE_IDLE) {
            mState = STATE_CREATED;
            if (DEBUG) {
                Log.d(TAG, "onCreate");
            }
            onCreate();
        }
    }

    public final void performStart() {
        if (mState == STATE_STARTED) {
            return;
        }
        if (mState == STATE_CREATED || mState == STATE_STOPPED) {
            mState = STATE_STARTED;
            if (DEBUG) {
                Log.d(TAG, "onStart");
            }
            onStart();
        }
    }

    public final void performStop() {
        if (mState == STATE_STOPPED) {
            return;
        }
        if (mState == STATE_STARTED) {
            mState = STATE_STOPPED;
            if (DEBUG) {
                Log.d(TAG, "onStop");
            }
            onStop();
        }
    }

    public final void performDestroy() {
        if (mState == STATE_DESTROYED) {
            return;
        }
        if (DEBUG) {
            Log.d(TAG, "onDestroy");
        }
        mState = STATE_DESTROYED;
        super.release();
        onDestroy();
    }

    public final boolean performBackPressed() {
        if (mState == STATE_DESTROYED) {
            return false;
        }
        if (DEBUG) {
            Log.d(TAG, "onBackPressed");
        }
        return onBackPressed();
    }

    public RouteComponentController(@Nullable IEventController eventController) {
        super(eventController);
    }
}

File: framework/src/main/java/com/thornbirds/frameworkext/component/route/RouteComponentController.kt
package com.thornbirds.frameworkext.component.route

import android.util.Log
import com.thornbirds.component.ComponentController
import com.thornbirds.component.IEventController
import com.thornbirds.component.utils.DebugUtils

/**
 * Created by yangli on 2017/11/19.
 */
internal typealias RouterController = RouteComponentController<IPageRouter<*>>
internal typealias PageEntry = IPageEntry<RouterController>
internal typealias PageCreator = IPageCreator<PageEntry>

abstract class RouteComponentController<T : IRouter<*>>(
    eventController: IEventController?
) : ComponentController(eventController), IRouterController {

    abstract val router: T?
    abstract val parentRouter: T?

    abstract fun exitRouter(): Boolean

    var state = STATE_IDLE
        private set

    protected abstract fun onCreate()
    protected abstract fun onStart()
    protected abstract fun onStop()
    protected abstract fun onDestroy()
    protected abstract fun onBackPressed(): Boolean

    fun performCreate() {
        if (state == STATE_CREATED) {
            return
        }
        if (state == STATE_IDLE) {
            state = STATE_CREATED
            if (DebugUtils.DEBUG) {
                Log.d(TAG, "onCreate")
            }
            onCreate()
        }
    }

    fun performStart() {
        if (state == STATE_STARTED) {
            return
        }
        if (state == STATE_CREATED || state == STATE_STOPPED) {
            state = STATE_STARTED
            if (DebugUtils.DEBUG) {
                Log.d(TAG, "onStart")
            }
            onStart()
        }
    }

    fun performStop() {
        if (state == STATE_STOPPED) {
            return
        }
        if (state == STATE_STARTED) {
            state = STATE_STOPPED
            if (DebugUtils.DEBUG) {
                Log.d(TAG, "onStop")
            }
            onStop()
        }
    }

    fun performDestroy() {
        if (state == STATE_DESTROYED) {
            return
        }
        if (DebugUtils.DEBUG) {
            Log.d(TAG, "onDestroy")
        }
        state = STATE_DESTROYED
        super.release()
        onDestroy()
    }

    fun performBackPressed(): Boolean {
        if (state == STATE_DESTROYED) {
            return false
        }
        if (DebugUtils.DEBUG) {
            Log.d(TAG, "onBackPressed")
        }
        return onBackPressed()
    }

    companion object {
        const val STATE_IDLE = 0
        const val STATE_CREATED = 1
        const val STATE_STARTED = 2
        const val STATE_STOPPED = 3
        const val STATE_DESTROYED = 4
    }
}